File: 224_basic_calculator/calculate_v2.cpp
#include <iostream>
#include <string>
#include <stack>
#include <vector>

using namespace std;

bool isdigit(char c) {
    return ( '0' <= c && c <= '9' );
}

class Solution {
    public:
        int calculate(string s) {
            stack<int> vals;
            stack<char> ops;

            for(int i = 0; i < s.size(); i++) {
                if(s[i] == '(') continue;
                else if(s[i] == '+' || s[i] == '*') ops.push(s[i]);
                else if(s[i] == ')') {
                    char op = ops.top();
                    ops.pop();
                    if(op == '+') {
                        int val = vals.top();
                        vals.pop();
                        val += vals.top();
                        vals.pop();
                        vals.push();
                    }
                }
            }
        }
};

int main() {
    string s = "(1+(4+5+2)-3)+(6+8)";
    Solution sol;
    cout << sol.calculate(s) << endl;

}

File: 224_basic_calculator/calculate_v2.py
class Solution(object):
    def calculate(self, s):
        s = '(' + s + ')'
        ops = []
        vals = []
        i = 0
        while i < len(s):
            print vals, ops
            if s[i]  == ' ': 
                i += 1
                continue
            elif s[i] in {'+', '-', '('}:
                ops.append(s[i])
            elif s[i] == ')':
                while ops[-1] != '(':
                    op = ops.pop()
                    if op == '+': vals.append(vals.pop() + vals.pop())
                    elif op == '-': vals.append( (-1) * vals.pop() + vals.pop())
                ops.pop()
            else:
                j = i
                while j < len(s) and s[j].isdigit(): j += 1
                vals.append(int(s[i:j]))
                i = j
                continue
            i += 1

        return vals[-1]


sol = Solution()
s = " 2-1 + 2 "
print s
print sol.calculate(s)
    	
